feat(run-mount): add support for variable expansion

omermizr opened the pull request:
This PR adds support for variable expansion in RUN mounts.
Resolves https://github.com/moby/buildkit/issues/815

Caveats:
- `from` expansion is not supported.
- Expansion is not supported in `ONBUILD RUN` commands (I don't even think mounts work there).

omermizr commented:
@tonistiigi FYI

tonistiigi commented on frontend/dockerfile/dockerfile2llb/convert_runmount.go:
```diff
@@ -15,33 +15,36 @@ import (
+	for i, mount := range mounts {
+		var from string
+		if mount.From == "" {
+			// this might not be accurate because the type might not have a real source (tmpfs for instance),
```
I don't understand why this doesn't check for type anymore.

omermizr replied:
If we want to be able to expand `type`, we can't check for it here. Fortunately (AFAICT), this function is responsible for getting mount sources so they can be registered, so we can use a fallback value (`emptyImageName` in this case) as a source, which again, AFAICT should be fine. Do you see any problem with that?

tonistiigi commented on frontend/dockerfile/instructions/commands_runmount.go:
```diff
@@ -64,13 +65,25 @@ func runMountPreHook(cmd *RunCommand, req parseRequest) error {
+	return setMountState(cmd, false, nil)
+}
+
+func getMountState(cmd *RunCommand) *mountState {
```
why has this moved?

tonistiigi commented on frontend/dockerfile/instructions/commands.go:
```diff
@@ -272,6 +272,13 @@ type RunCommand struct {
 }
 
+func (c *RunCommand) Expand(expander SingleWordExpander) error {
+	if err := SetMountState(c, expander); err != nil {
```
don't make this public and call it `expandMounts`.

omermizr replied:
sure thing, but given that this function actually sets the mounts, doesn't `setMountState` make more sense?

tonistiigi commented on frontend/dockerfile/instructions/commands_runmount.go:
```diff
@@ -80,12 +93,8 @@ func runMountPostHook(cmd *RunCommand, req parseRequest) error {
-	}
-	return v.(*mountState)
+func SetMountState(cmd *RunCommand, expander SingleWordExpander) error {
+	return setMountState(cmd, true, expander)
```
don't think you need second param, just check nil.

tonistiigi commented on frontend/dockerfile/instructions/commands_runmount.go:
```diff
@@ -151,6 +160,25 @@ func parseMount(value string) (*Mount, error) {
 		}
 
 		value := parts[1]
+		// possible env var, defer for later
```
what case is this?

omermizr replied:
`parseMount` is called twice - once when the command is initially parsed, and then again when dispatching the command. For the first call, we don't have a way to expand variables.

tonistiigi replied:
Why do we need to check for the case? Can't we just check `expander==nil`, then expand the value if set, if `from` changes then show an error.

tonistiigi replied:
Note that linter also complains about this.

omermizr replied:
I'm not sure I understand. 
We check `expander == nil` and expand the value if set. Right now, if the `from` value has a `$` I return an informative error. Anyway, this code changed a bit so make sure you're looking at the most recent version

tonistiigi replied:
What I'm saying is that I don't think we need to check if the value contains a variable. It's redundant (and possibly could diverge from expander implementation). We can just call expander and make decisions based on the return value.

omermizr replied:
Gotcha. Will do that, but that means the failure will happen in the dispatch stage rather than the parsing stage

tonistiigi replied:
That's fine. Don't forget to squash commits as well.

omermizr replied:
I implemented this logic, but then I noticed it fails long before we get to `dispatch`. 
Basically in the case that `from` is a variable, a new stage is added and its base name is invalid.
See `frontend/dockerfile/dockerfile2llb/convert.go:233`.
How should we proceed?

tonistiigi replied:
I don't get it. Vars are still not allowed in `from` and should produce an error. (Could actually add a test for that). But to check that from doesn't contain a variable you don't need special duplicate logic but can just compare the value of the expander.

omermizr replied:
The problem is that this isn't supported by the current flow:
1) Initial parsing. In this stage we don't have the `expander`, so we can't check the diff as suggested.
2) Processing the build stages and validation. Here, if a user used a variable in `from`, an error will be produced because of the stages validation (basically, a build stage name can't be uppercase or contain a `$`).
3) Dispatching commands. Here we can compare the expanded value to the original one, but we won't even get here because of the previous stage.

I hope that makes sense :)

tonistiigi replied:
Ok, leave the validation for `from` to get a clearer error message. The others don't need it and can rely only on the expander implementation.

omermizr replied:
Done. 
It came out a bit verbose, but it was either that or extracting this whole block to a separate function with some weird inputs/outputs.
I also thought about using an anonymous function so I could use closures, but I wasn't sure if that would be in the spirit of Go.
LMK if you have a better idea of if this is good enough

tonistiigi commented on frontend/dockerfile/instructions/commands_runmount.go:
```diff
@@ -151,6 +160,25 @@ func parseMount(value string) (*Mount, error) {
+		if err != nil {
+			return nil, err
+		}
+		if matched {
```
Not all variables are allowed to be expended (atm). From/type are not allowed afaics. Add helpful errors to these cases.

omermizr replied:
Added message for `from`. Regarding `type` - let's keep discussing in the other comment

tonistiigi reviewed (commented):
Missing DCO

I don't fully get the changes in `convert.go`. If it is just refactoring with no functional changes it should be in a different PR.

This doesn't fix the issues with output afaics. Fine by me if you want to tackle this in follow-up but should be fixed as well.

omermizr commented:
Added DCOs.
Removed changes in `convert.go`.
Regarding output formatting - I think it's beyond the scope of this PR, especially considering it's an existing problem. I also don't wanna mix a feature branch with a fix branch. I can make a separate PR to fix it, but I couldn't figure out where that code is. If you can send me on a path that'd be great :)

tonistiigi commented on frontend/dockerfile/dockerfile_mount_test.go:
```diff
@@ -221,3 +225,126 @@ RUN --mount=type=cache,target=/mycache [ -f /mycache/foo ]
+ENV MNT_ID=mycache
+ENV MNT_TYPE2=cache
+RUN --mount=type=$MNT_TYPE2,id=$MNT_ID,target=/abcabc touch /abcabc/foo
+RUN --mount=type=$MNT_TYPE2,id=$MNT_ID,target=/cbacba [ -f /cbacba/foo ]
```
This test should check different values. Currently for example if expansion doesn't work and always returns the same constant value, this test would still pass.

omermizr replied:
Thanks, good catch

tonistiigi reviewed (commented):
Only small comments left. Make sure to also squash commits and verify CI

tonistiigi commented on frontend/dockerfile/instructions/commands_runmount.go:
```diff
@@ -151,13 +156,29 @@ func parseMount(value string) (*Mount, error) {
 
 		value := parts[1]
+		// check for potential variable
+		if expander != nil {
```
You can just do the `from` check in here without requiring per-key checks. I think we should also not try at expansion at all for fields we ignore the error. It's kind of a weird logic where we call function knowing it will fail. Unless we do it in some other way we should not attempt on the first try. The fields except `from` are not allowed to be used after first pass with this logic. With ignoring the errors this might go unnoticed.

DCO+squash

omermizr replied:
This is actually what I did initially, but I thought you'd prefer to handle it on the key level. Anyway, I'll push another iteration in a few

tonistiigi commented on frontend/dockerfile/instructions/commands_runmount.go:
```diff
@@ -151,6 +156,23 @@ func parseMount(value string) (*Mount, error) {
+			if matched, err := regexp.MatchString(`\$.`, value); err != nil { //nolint
+				return nil, err
+			} else if matched {
+				return nil, errors.Errorf("'%s' doesn't support variable expansion", key)
```
nit: add ", define alias stage instead" to the error message
